Add Lambda function & build + MerkleTree variant that splits memory between disk and memory

rvagg opened the pull request:
caveat emptor, this is some cowboy stuff from a not very experienced Rust programmer, but it seems to get the right results with my test data and I can squeeze it into Lambda's memory.

rvagg committed:
added lambda compile option incl docker and Makefile

rvagg committed:
implement MerkleTree caching store that splits disk and memory

rvagg commented:
With a plain `VecStore`, a 1G piece will take up just under 3,100Mb to calculate with the combination of the padded version and Merkle caching in the `VecStore`.

Lambda has a limit of 3,008Mb in memory and 500Mb of disk. This includes whatever additional is needed to run the executable and whatever dynamic libraries it needs to pull in. So we're too big in memory.

This PR includes a `MultiStore` that uses around 375Mb in a `DiskStore` and the rest in a `VecStore`, so we can make use of that scratch disk space to scrape under the limit.

rvagg commented:
The storage stuff it's implementing is in https://github.com/filecoin-project/merkle_light/tree/v0.15.2/src/store FYI

vmx commented on src/commp.rs:
```diff
@@ -0,0 +1,261 @@
+    let padded_size = pad_reader.padsize;
+
+    // Grow the vector big enough so that it doesn't grow it automatically
+    let mut data = Vec::with_capacity((padded_size as u64 as f64 * 1.01) as usize);
```
I stumbled upon https://crates.io/crates/arrayvec recently. I haven't used it and can't tell about performance characteristics or if it even works with such large allocations. But it might be worth a try for having full control over the memory.

vmx commented on Cargo.toml:
```diff
@@ -2,11 +2,28 @@
+serde_json = "^1"
+serde_derive = "^1"
+log = "^0.4"
+simple_logger = "^1"
```
FYI: I've had good experience with https://crates.io/crates/flexi_logger so far.

rvagg committed:
fixup! implement MerkleTree caching store that splits disk and memory

vmx commented on src/commp.rs:
```diff
@@ -0,0 +1,261 @@
+    return pad_reader;
+}
+
+#[allow(dead_code)]
```
Not sure which parts are dead code here, but generally allow that sounds concerning :)

rvagg replied:
there's 2 bins in here, `commp` (main.rs) and `bootstrap` (lambda.rs), both use commp.rs but lambda.rs only uses one of the exposed `pub fn`s and without these `dead_code` macros I get compiler warnings when compiling `bootstrap`. Is there a better way of handling this? It does seem a bit silly.

vmx replied:
When this is done as a library (as #4 does) then it isn't an issue anymore as `pub` functions in a library are not considered dead code.

vmx commented on src/commp.rs:
```diff
@@ -0,0 +1,261 @@
+use storage_proofs::pieces::generate_piece_commitment_bytes_from_source;
+use storage_proofs::util::NODE_SIZE;
+
+#[path = "multistore.rs"]
```
That derive shouldn't be needed.

rvagg replied:
If I remove `#[path = "multistore.rs"]` it won't compile. I can't figure out how to make this thing exist in its own file without this pragma or doing the whole mod directory thing (which I still don't understand either)

vmx replied:
I had to check myself and ended up with a PR which illustrates what the problem was: https://github.com/rvagg/rust-fil-commp-generate/pull/4

Generally spoken: In Rust you have either a library or a single binary application. In your case, with multiple binaries, you'd like to have your module as a library, hence your entry point is in `src/lib.rs`. Often this file is pretty much empty and just includes the other files/modules. To make sure the other modules are found, you need to reference them from `src/lib.rs` with `mod`.

You won't use `mod` within your binaries, you will just import things via `use` there.

vmx commented on src/commp.rs:
```diff
@@ -0,0 +1,261 @@
+          print!("reached file size, now padding ...")
+        }
+        */
+        if self.pos >= self.size {
```
You could do the `self.pos ...` part outside of the if-else and return the `cs` directly. So it would be roughly:

```rust
let cs = if …{
    …
    cmp::min(…)
} else {
    self.inp.read(buf)?
}
self.post += cs
Ok(cs)
```

vmx commented on src/commp.rs:
```diff
@@ -0,0 +1,261 @@
+// figure out how big this piece will be when padded
+fn padded_size(size: u64) -> u64 {
+    let logv = 64 - size.leading_zeros();
+    let sect_size = (1 as u64) << logv;
```
`1 as u64` can be written as `1u64`.

vmx commented on src/commp.rs:
```diff
@@ -0,0 +1,261 @@
+    let comm_p = tree.root();
+    comm_p
+        .write_bytes(&mut comm_p_bytes)
+        .expect("borked at extracting commp bytes");
```
`expect()` return leads to a panic. What you could do here as you return an "`anyhow` Result" is using `context(…)` instead. This chains up anyhow error messages and you get a some nice output if it fails.

I won't mention that at further occurrences again.

rvagg replied:
I can't work out how to use this in my `generate_commp_filecoin_proofs()` implementation which returns a plain `Result` which comes out of `generate_piece_commitment()`. Doing a `context()` with that gives me `the trait std::convert::From<anyhow::Error> is not implemented for std::io::Error` which I suppose means i need to wrap it somehow but can't figure that out. So I still have that `expect()` in place. Let me know if you have a better idea for it.

vmx replied:
You would need to change the return value from `Result<CommP, std::io::Error>`to `Result<CommP, anyhow::Error>`. The `generate_piece_commitment()` call will still return a `std::io::Error`, but the `context()` call with then take that error and wrap it in an `anyhow::Errror` with the additional information that you provided in the context.

Error handling can be tricky, I recommend reading https://blog.burntsushi.net/rust-error-handling/ it's long but totally worth it.

vmx commented on src/commp.rs:
```diff
@@ -0,0 +1,261 @@
+        inp: inp,
+    };
+
+    return pad_reader;
```
`return` statement is not needed, you can use an implicit return here, so it's just  `pad_reader` without the semi-colon at the end.

vmx commented on src/commp.rs:
```diff
@@ -0,0 +1,261 @@
+    size: u64,
+) -> Result<CommP, std::io::Error>
+where
+    R: io::Read,
```
I think (not sure though) that you can write `R: io::Read + Sized` here and using just `generate_commp_filecoin_proofs<R>()`. This way you have all the bounds requirements in one place.

vmx commented on src/commp.rs:
```diff
@@ -0,0 +1,261 @@
+            &mut temp_piece_file,
+            PaddedBytesAmount::from(piece_size).into(),
+        )
+        .unwrap();
```
As this function returns a result, it's cleaner to just propagate the error up via `?` instead of panicking via `unwrap()`.

rvagg replied:
same thing as the previous `Result` garbage: `the trait std::convert::From<anyhow::Error> is not implemented for std::io::Error`, I have no idea.

vmx commented on src/commp.rs:
```diff
@@ -0,0 +1,261 @@
+            &mut temp_piece_file,
+            PaddedBytesAmount::from(piece_size).into(),
+        )
+        .unwrap();
```
Same as above.

vmx commented on src/commp.rs:
```diff
@@ -0,0 +1,261 @@
+    temp_piece_file.seek(SeekFrom::Start(0))?;
+    let commitment: Fr32Ary;
+
+    if multistore {
```
You can use implicit returns in if statements. So this would be:

```rust
let commitment = if multistore {
    local_multistore_generate_piece_commitment_bytes_from_source::<
            DefaultPieceHasher,
            Cursor<&mut Vec<u8>>,
        >(
            &mut temp_piece_file,
            PaddedBytesAmount::from(piece_size).into(),
        )?
} else {
…
};
```

vmx commented on src/lambda.rs:
```diff
@@ -0,0 +1,94 @@
+use std::error::Error;
+use std::str::FromStr;
+
+#[macro_use]
```
In Rust editiion 2018, you can import macros directly using normal imports. So you don't need to use `#[macro_use]` or `extern crate` anymore. I guess the `lamda!` macro comes from this crate, so it would just be:

```rust
use lambda_runtime::lamda
```

in order to import the lamnda macro.

vmx commented on src/lambda.rs:
```diff
@@ -0,0 +1,94 @@
+#[macro_use]
+extern crate serde_derive;
+#[macro_use]
+extern crate log;
```
The same as above also for the logging. Within the module you use logging, you just import the macros directly:

```rust
use log::{debug, info}
```

vmx commented on src/lambda.rs:
```diff
@@ -0,0 +1,94 @@
+    let result = client
+        .get_object(get_req)
+        .sync()
+        .expect("Couldn't GET object");
```
I think here it's OK, but generally spoken it's better to propagate errors up rather than panicking.

rvagg replied:
easier said than done! error handling and propagation is one thing I can't get a grip of yet, this use of `anyhow` seems to complicate things (I'm only pulling it in because some of rust-fil-proofs dabbles with it). `expect`, `unwrap`, `?`, are still fairly opaque to me in where they should, can and can't be used.

vmx commented on src/lambda.rs:
```diff
@@ -0,0 +1,94 @@
+extern crate serde_derive;
+#[macro_use]
+extern crate log;
+extern crate simple_logger;
```
Not needed, just remove it as you cal use top-level modules without importing them in the Rust 2018 edition.

vmx commented on src/main.rs:
```diff
@@ -2,164 +2,44 @@
 
-use anyhow::ensure;
+#[macro_use]
+extern crate log;
```
Remove those to lines and instead import the logging macros directly in the modules where you use them (`use log::{debug, info}`).

vmx commented on src/main.rs:
```diff
@@ -2,164 +2,44 @@
-
-    Ok(comm_p_bytes)
-}
+extern crate simple_logger;
```
Not needed can simply be removed.

rvagg replied:
why do I still need these bits:

```
#[macro_use]
extern crate lambda_runtime as lambda;
#[macro_use]
extern crate serde_derive;
```
?
is `log` set up different such that `info!()` and friends are exported in some special way and the serde and lambda macros aren't?

rvagg replied:
nevermind, figured it out:

```
use lambda_runtime::error::HandlerError;
use lambda_runtime::{lambda, Context};
use serde_derive::{Deserialize, Serialize};
```

vmx commented on src/main.rs:
```diff
@@ -2,164 +2,44 @@
     }
     let filename = &args[2];
-    let file = File::open(filename).expect("Unable to open file");
+    let mut file = File::open(filename).expect("Unable to open file");
```
As your main function already returns a result, you can return an error here instead of panicking:

```rust
let mut file = std::fs::File::open(filename).map_err(|_| format!("Unable to open file: {}", filename))?;
```
or if returning the original error (`Os { code: 2, kind: NotFound, message: "No such file or directory" }`) is good enough simply

```rust
let mut file = File::open(filename)?;
```

vmx commented on src/multistore.rs:
```diff
@@ -0,0 +1,131 @@
+use merkletree::store::{DiskStore, Store, StoreConfig, VecStore};
+
+#[derive(Debug)]
+pub struct MultiStore<E: Element>(DiskStore<E>, VecStore<E>);
```
I would make this a struct with fields:

```rust
pub struct MultiStore<E: Element> {
    disk: DiskStore<E>,
    mem: VecStore<E>
};
```

This way you can access them with `self.disk` and `self.mem` instead of `self.0` and `self.1` which is easier to read and less error-prone.

rvagg replied:
nice!

vmx commented on src/multistore.rs:
```diff
@@ -0,0 +1,131 @@
+                    start: 0,
+                    end: r.end - DISK_MAX,
+                };
+                let r0 = self.1.read_range(nr0).unwrap();
```
Shouldn't this be `self.0` as the beginning of the range is on the disk part?

rvagg replied:
yes, you're right, and this is concerning since I was getting the right numbers out of it! now I have to figure out how, because I'm pretty sure split ranges were involved in my test data.

vmx commented on src/multistore.rs:
```diff
@@ -0,0 +1,131 @@
+                end: r.end - DISK_MAX,
+            };
+            self.1.read_range(nr)
+        } else {
```
This is just a personal taste thing. I would make it an `else if` as it is then clearly split into three parts:

 1. Fully in memory
 2. First few bytes on disks, rest on memory 
 3. All on disk

vmx reviewed (commented):
I've only reviewed the Rust code and not the other code like scripts or Docker files.

vmx commented:
I should make clear that almost everything from the review is just general notes about the Rust code in order to help @rvagg learning Rust. I found only a single thing that I think is a bug: https://github.com/rvagg/rust-fil-commp-generate/pull/3#discussion_r379408702.

rvagg committed:
fixup! implement MerkleTree caching store that splits disk and memory

rvagg committed:
fixup! implement MerkleTree caching store that splits disk and memory

vmx commented on src/multistore.rs:
```diff
@@ -74,35 +77,36 @@ impl<E: Element> Store<E> for MultiStore<E> {
+                start: 0,
+                end: r.end - DISK_MAX,
+            };
+            let rdisk = self.mem.read_range(nrdisk).unwrap();
```
Have you already figured out why this works as I still think it shouldn't? Do your tests break when you change it to `self.disk.read_range(nrdisk)`?

rvagg replied:
I think I've got it now ... I was getting the same result commp regardless of whether this was switched around. My initial version of this had a panic when you queried a split range because <too hard> but when I ran it, it panicked, so I had to implement it! So those two facts combined had me confused and worried because it seemed like it didn't matter what data was being fed.

However, I've gone over the code again and I _think_ I had `>` and `>=` wrongly used, I also implemented the split copy in `copy_from_slice` which is essentially the write version of this `range_read` (see 6ae7072). But now I have logging in both of those split blocks and it's no longer logging. I can't recall where I originally got the size of the `DISK_MAX` from, `262144 * 48` but I think 262144 was from inspecting how it worked originally and it seeming to work on 262144 chunks. Multiplied by 48 gave me roughly how much disk I wanted to use up. So 🤞 I think I'm hitting right on aligned slices and not needing these split read/write operations anymore. Hopefully my code now would survive if it did, though, but it's another reason we need to do a lot of testing of this stuff.

vmx committed:
feat: move binaries source into the `src/bin` directory

This project is now a library which is used by two binaries named
`commp` and `lambda`.

vmx reviewed (commented):
I think I've addressed all your questions, in case I've missed anything or there are new ones, please let me know.

rvagg committed:
fixup! implement MerkleTree caching store that splits disk and memory

rvagg committed:
fixup! feat: move binaries source into the `src/bin` directory

rvagg committed:
fixup! implement MerkleTree caching store that splits disk and memory

rvagg commented:
merged @ 98a1c7ff7036261095a82fe7a8ec7645e31c86c3
